lib/search.js

```diff
@@ -0,0 +1,166 @@
+/**
+ * Copyright 2015-2016, Google, Inc.
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+'use strict';
+
+const URL = require('url');
+const elasticlunr = require('elasticlunr');
+const libCache = require('../lib/data-cache');
+const libPwa = require('../lib/pwa');
+
+const CACHE_LIFETIME = 60 * 60 * 24 * 7; // 7 days
+const SEARCH_INDEX_CHANGE = 'SearchIndexChange';
+
+/**
+ * Search class for the elasticlunr functions
+ *
+ * Exports a singleton object instance
+ */
+class Search {
+
+  constructor() {
+    this._initIndex();
+  }
+
+  _initIndex() {
+    this._index = elasticlunr(function() {
+      this.setRef('id');
+      this.addField('displayName');
+      this.addField('urlText');
+    });
+    this._modified = new Date();
+  }
+
+  /**
+   * Create a doc element from a PWA.
+   *
+   * @param {PWA} PWA to index
+   * @return {doc} a doc for the text search engine
+   */
+  _docFromPwa(pwa) {
+    const url = URL.parse(pwa.absoluteStartUrl);
+    const urlText = url.hostname.replace(/\.|-/g, ' ');
+    return {
+      id: pwa.id,
+      displayName: pwa.displayName,
+      urlText: urlText
+    };
+  }
+
+  /**
+   * Add a PWA to the search index.
+   *
+   * @param {PWA} PWA to index
+   * @return {Promise<doc>} the doc added to the text search engine
+   */
+  addPwa(pwa) {
+    const doc = this._docFromPwa(pwa);
+    this._index.addDoc(doc);
+    this.sarchIndexChange();
+    return Promise.resolve(doc);
+  }
+
+  addPwas(pwas) {
+    pwas.forEach(pwa => {
+      const doc = this._docFromPwa(pwa);
```

Comment:
The next two lines are repeated from the addPwa method. Maybe extracting to another method and using it in both methods?